### ISSUE TITLE
False positive for "clippy::needless_collect"

### ISSUE THREAD

**Issue by porky11** [rust-lang/rust-clippy#13430]
### Summary

I'm creating a vector from an iterator using `collect`, since I want to iterate over some iterator multiple times.igina
But the original iterator didn't implement `Clone`, so I had to collect the elements into a vector first, so I could get an iterator which implements `Clone`.

In my case the iterator was an iterator over the lines of a file using `BufReader::lines`.
But I tried to find out the exact cause of the issue and generalized/simplified the issue in the example.

### Lint Name

clippy::needless_collect

### Reproducer

I tried this code:

```rust
fn test(elements: impl Iterator<Item = Option<usize>>) {
    let elements: Vec<_> = elements.into_iter().map(Option::unwrap).collect();
    let elements = elements.into_iter();

    for _element in elements.clone() {
        // do something with element
    }

    for _element in elements {
        // do something with element
    }
}
```

I got this message:

```
warning: avoid using `collect()` when not needed
 --> src/main.rs:6:69
  |
6 |     let elements: Vec<_> = elements.into_iter().map(Option::unwrap).collect();
  |                                                                     ^^^^^^^
7 |     let elements = elements.into_iter();
  |                    -------------------- the iterator could be used here instead
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect
  = note: `-W clippy::needless-collect` implied by `-W clippy::nursery`
  = help: to override `-W clippy::nursery` add `#[allow(clippy::needless_collect)]`
help: use the original Iterator instead of collecting it and then producing a new one
  |
6 ~     
7 ~     let elements = elements.into_iter().map(Option::unwrap);
  |

```

If I follow this suggestion, my code looks like this:

```rust
fn test(elements: impl Iterator<Item = Option<usize>>) {
    let mut elements = elements.map(Option::unwrap);

    for _element in elements.clone() {
        // do something with element
    }

    for _element in elements {
        // do something with element
    }
}
```

But following this suggestion caused me to get another error:

```
error[E0599]: the method `clone` exists for struct `Map<impl Iterator<Item = Option<usize>>, fn(Option<usize>) -> usize {Option::<usize>::unwrap}>`, but its trait bounds were not satisfied
  --> src/main.rs:8:30
   |
8  |     for _element in elements.clone() {
   |                              ^^^^^
   |
  ::: /home/p11/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/adapters/map.rs:62:1
   |
62 | pub struct Map<I, F> {
   | -------------------- doesn't satisfy `_: Clone`
   |
   = note: the following trait bounds were not satisfied:
           `impl Iterator<Item = Option<usize>>: std::clone::Clone`
           which is required by `std::iter::Map<impl Iterator<Item = Option<usize>>, fn(std::option::Option<usize>) -> usize {std::option::Option::<usize>::unwrap}>: std::clone::Clone`
help: consider restricting the type parameter to satisfy the trait bound
   |
5  | fn test(elements: impl Iterator<Item = Option<usize>>) where impl Iterator<Item = Option<usize>>: std::clone::Clone {
   |                                                        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

This suggestion also isn't helpful, since I can't turn my iterator into one which implements `Clone`.

But I already found a solution which works while writing this:

```rust
fn test(elements: impl Iterator<Item = Option<usize>>) {
    let mut elements = elements.map(Option::unwrap);

    for _element in elements.by_ref() {
        // do something with element
    }

    for _element in elements {
        // do something with element
    }
}
```

Not sure if this solution always works and doesn't cause any other issues.

### Version

rustc 1.80.1 (3f5fd8dd4 2024-08-06)
binary: rustc
commit-hash: 3f5fd8dd41153bc5fdca9427e9e05be2c767ba23
commit-date: 2024-08-06
host: x86_64-unknown-linux-gnu
release: 1.80.1
LLVM version: 18.1.7

**Comment by ronnodas** [comment]
Your workaround "solution" using `by_ref` has different semantics: the second for loop resumes where the first one ends, so will run 0 times if the first one does not encounter a `break`.

**Comment by porky11** [comment]
@ronnodas Thanks, you're right.
